test/hotspot/jtreg/compiler/lib/verify/Verify.java

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.lib.verify;
+
+import java.util.Optional;
+import java.lang.foreign.*;
+
+/**
+ * The Verify class provides a single Verify.checkEQ static method, which recursively compares
+ * the two Objects by value. It deconstructs Object[], compares boxed primitive types, and
+ * compares the content of arrays and MemorySegments.
+ *
+ * When a comparison fail, then methods print helpful messages, before throwing a VerifyException.
+ */
+public final class Verify {
+    /**
+     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     */
+    public static void checkEQ(Object a, Object b) {
+        checkEQ(a, b, "");
+    }
+
+    /**
+     * Verify the content of two Objects, possibly recursively. Only limited types are implemented.
+     */
+    private static void checkEQ(Object a, Object b, String context) {
+        // Both null
+        if (a == null && b == null) {
+            return;
+        }
+
+        // Null mismatch
+        if (a == null || b == null) {
+            System.err.println("ERROR: Verify.checkEQ failed: null mismatch");
+            print(a, "a " + context);
+            print(b, "b " + context);
+            throw new VerifyException("Object array null mismatch.");
```

Comment:
Why does this message mention "array"? From what I understand, it can compare any two objects.

Reply:
Good catch. I used to only have it work for arrays ;)